### PR TITLE
apc_modbus: Use snprintf instead of strncpy

### LINKED ISSUE #2122: apc_modbus: truncation compiler warning
Good morning!

@EchterAgo & @jimklimov & others :
**THANK YOU** for the immense efforts with `apc_modbus` - great to have this included in NUT.

Building on my Slackware 15.0/gcc11.2.0 environment:
```
In function â€˜_apc_modbus_parse_host_portâ€™,
    inlined from â€˜upsdrv_initupsâ€™ at apc_modbus.c:1246:8:
apc_modbus.c:1137:17: warning: â€˜strncpyâ€™ specified bound 256 equals destination size [-Wstringop-truncation]
 1137 |                 strncpy(host, start, host_buf_size);
      |                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
apc_modbus.c: At top level:
cc1: note: unrecognized command-line option â€˜-Wno-unknown-warning-optionâ€™ may have been intended to silence earlier diagnostics
cc1: note: unrecognized command-line option â€˜-Wno-reserved-identifierâ€™ may have been intended to silence earlier diagnostics
```
P.S.: building with the patched libmodbus (if that's of any relevance here)

### PR DESCRIPTION
**Pull request by EchterAgo**
`strncpy` does not necessarily zero terminate the buffer so we replace it with `snprintf` and add error checking.

This also fixes an off-by-one in the `snprintf` error checking of the `_apc_modbus_double_to_nut` and `_apc_modbus_power_to_nut` functions and adds error checking to every `snprintf` call.

fixes #2122 

## General points

- [x] Described the changes in the PR submission or a separate issue, e.g.
  known published or discovered protocols, applicable hardware (expected
  compatible and actually tested/developed against), limitations, etc.

- [x] There may be multiple commits in the PR, aligned and commented with
  a functional change. Notably, coding style changes better belong in a
  separate PR, but certainly in a dedicated commit to simplify reviews
  of "real" changes in the other commits. Similarly for typo fixes in
  comments or text documents.

## Frequent "underwater rocks" for driver addition/update PRs

- [x] Revised existing driver families and added a sub-driver if applicable
  (`nutdrv_qx`, `usbhid-ups`...) or added a brand new driver in the other
  case.

- [x] Did not extend obsoleted drivers with new hardware support features
  (notably `blazer` and other single-device family drivers for Qx protocols,
  except the new `nutdrv_qx` which should cover them all).

- [x] For updated existing device drivers, bumped the `DRIVER_VERSION` macro
  or its equivalent.

- [x] For USB devices (HID or not), revised that the driver uses unique
  VID/PID combinations, or raised discussions when this is not the case
  (several vendors do use same interface chips for unrelated protocols).

- [x] For new USB devices, built and committed the changes for the
  `scripts/upower/95-upower-hid.hwdb` file

- [x] Proposed NUT data mapping is aligned with existing `docs/nut-names.txt`
  file. If the device exposes useful data points not listed in the file, the
  `experimental.*` namespace can be used as documented there, and discussion
  should be raised on the NUT Developers mailing list to standardize the new
  concept.

- [x] Updated `data/driver.list.in` if applicable (new tested device info)

## Frequent "underwater rocks" for general C code PRs

- [x] Did not "blindly assume" default integer type sizes and value ranges,
  structure layout and alignment in memory, endianness (layout of bytes and
  bits in memory for multi-byte numeric types), or use of generic `int` where
  language or libraries dictate the use of `size_t` (or `ssize_t` sometimes).

- [x] Progress and errors are handled with `upsdebugx()`, `upslogx()`,
  `fatalx()` and related methods, not with direct `printf()` or `exit()`.
  Similarly, NUT helpers are used for error-checked memory allocation and
  string operations (except where customized error handling is needed,
  such as unlocking device ports, etc.)

- [x] Coding style (including whitespace for indentations) follows precedent
  in the code of the file, and examples/guide in `docs/developers.txt` file.

- [x] For newly added files, the `Makefile.am` recipes were updated and the
  `make distcheck` target passes.

## General documentation updates

- [x] Updated `docs/acknowledgements.txt` (for vendor-backed device support)

- [x] Added or updated manual page information in `docs/man/*.txt` files
  and corresponding recipe lists in `docs/man/Makefile.am` for new pages

- [x] Passed `make spellcheck`, updated spell-checking dictionary in the
  `docs/nut.dict` file if needed (did not remove any words -- the `make`
  rule printout in case of changes suggests how to maintain it).

## Additional work may be needed after posting this PR

- [ ] Propose a PR for NUT DDL with detailed device data dumps from tests
  against real hardware (the more models, the better).

- [ ] Address NUT CI farm build failures for the PR: testing on numerous
  platforms and toolkits can expose issues not seen on just one system.

- [ ] Revise suggestions from LGTM.COM analysis about "new issues" with
  the changed codebase.

### CODE REVIEW

**Comment by desertwitch** [comment]
Thanks a lot - can confirm this compiles without problems on my system now, let's see what the CIs do. :-)

File: drivers/apc_modbus.c [hunk 1]
```diff
@@ -1027,14 +1041,23 @@ static void _apc_modbus_usb_lib_to_nut(const modbus_usb_device_t *device, USBDev
...
+	res = snprintf(bus_buf, sizeof(bus_buf), "%03u", device->bus);
+	if (res < 0 || (size_t)res >= sizeof(bus_buf)) {
+		assert(0);
+	}
 	out->Bus = bus_buf;
 
-	snprintf(device_buf, sizeof(device_buf), "%03u", device->device_address);
+	res = snprintf(device_buf, sizeof(device_buf), "%03u", device->device_address);
+	if (res < 0 || (size_t)res >= sizeof(device_buf)) {
+		assert(0);
```
Comment (jimklimov) [inline drivers/apc_modbus.c:1052]: Is this supposed to error out and abort the driver? Typically we use a `fatalx(EXIT_FAILURE, "some explanation")` in these cases.

Comment (EchterAgo) [reply drivers/apc_modbus.c:1052]: Yes, it is supposed to do that, though maybe not a good idea if we ever decide to define `NDEBUG`. I'll make it a `fatalx`.

Comment (EchterAgo) [reply drivers/apc_modbus.c]: Updated.

**Pull request merged**